blog-post/blog-post.md

```diff
@@ -0,0 +1,382 @@
+As an example, consider the following code snippet that implements a transfer 
+function in Classic Solidity:
+
+```
+function transferTokens(
+    address from,
+    address to,
+    uint256 amount
+) public returns (bool) {
+    uint256 fromBalance = balances[from];
+    require(fromBalance >= amount, "Insufficient balance");
+    
+    balances[from] = fromBalance - amount;
+    balances[to] = balances[to] + amount;
+    
+    uint256 newFromBalance = balances[from];
+    uint256 newToBalance = balances[to];
+    
+    bool transferSuccess = (newFromBalance == fromBalance - amount) && 
+                          (newToBalance == balances[to] + amount);
+    
+    return transferSuccess;
+}
+```
+
+All local variables and function arguments and result needs to specify 
+its type. In Core Solidity, thanks to type inference, we could have
+completely omit type annotations, while keeping the guarantees about 
+the code type safety.
+
+```
+function transferTokens(from, to, amount) {
+    fromBalance = balances[from];
+    require(fromBalance >= amount);
+    
+    balances[from] = fromBalance - amount;
+    balances[to] = balances[to] + amount;
+    
+    newFromBalance = balances[from];
+    newToBalance = balances[to];
+    
+    transferSuccess = (newFromBalance == fromBalance - amount) && 
+                      (newToBalance == balances[to] + amount);
+}
+```
+
+## Extended example: Classic Solidity vs Core Solidity 
+
+Now, let's consider an extended example: a contract which implements a unified payment 
+processor that handles three different token standards. The complete Classic Solidity 
+implementation for this simple contract can be found [here.](https://gist.github.com/rodrigogribeiro/7cee270702987fe123921fdbbc2f12ff)
+The code starts by defining a `Payment` struct attempts to represent all three token standards, 
+using the `paymentType` field as a runtime discriminator to determine which fields are relevant 
+for each case.
+
+```
+enum PaymentType { NATIVE, ERC20, ERC721 }
+
+struct Payment {
+    PaymentType paymentType;  // Runtime type tag
+    address token;           // Used for ERC20/ERC721, must be zero for Native 
+    address from;            // Sender address  
+    address to;              // Receiver address
+    uint256 amount;          // Used for Native/ERC20, must be 1 for ERC721
+    uint256 tokenId;         // Used for ERC721, must be zero for others
+}
+```
+
+Next, functions `processPayment` and `calculateFee` uses explicit if/else chains 
+to manually check the `paymentType` and route to the appropriate handling logic.
+Also, in order to avoid invalid states, the code use `require` to ensure that 
+each case is dealing with a proper payment state value.
+
+```
+function processPayment(Payment calldata payment) external {
+  if (payment.paymentType == PaymentType.NATIVE) {
+    require(payment.token == address(0), "Native: no token");
+    require(payment.amount > 0, "Native: amount required");
```

Comment:
The Core Solidity version allows `amount == 0`. You should either allow it here or add the same check there. Otherwise the examples are not fully equivalent.

Same for `ERC20`.

Reply:
this section is gone now